Ask for python-cryptography v1.7+ in warnings and documentation

mtury opened the pull request:
I came across two undesired warnings from `ipsec.py`.

The first one was caused by the absence of `pycrypto`, which we don't use anymore.

The second one was caused by the location of `ModeWithAuthenticationTag` inside `cryptography`. It was moved from `interfaces` to `primitives.ciphers`, but this change does not seem to have been packaged in the latest stable Debian release.

gpotter2 commented:
I don't really know if this is wanted or not, but if you really want to remove that, you should update the doc as well:
https://github.com/secdev/scapy/blob/master/doc/scapy/installation.rst#optional-software-for-special-features

> IPsec Crypto Support. SecurityAssociation() needs Pycrypto 2.7a1. Combined AEAD modes such as GCM and CCM are not available yet because of cryptography restrictions.

mtury commented:
@ThomasFaivre Do you know the minimal version of `cryptography` needed for IPsec crypto support?

ThomasFaivre commented:
Hello!
I've developed under cryptography 1.4, but I've just passed the tests using 1.2
I did not manage to install cryptography versions under 1.2 with openssl-1.0.1t on debian-8, nor with openssl-1.0.1e on debian-7.
I think it could be a good thing to impose a cryptography version above 1.2 as there is a backward incompatible change in the certificate extensions in this version. If we want to update the TLS layer to use cryptography, it could be a thing to have a stable API. Also, if version below 1.2 are not compiling anymore, it a good reason not to use them!
I will also add that I advise to always use the latest cryptography as there might be security fixes like in version 1.5.3.

Changelog: https://cryptography.io/en/latest/changelog/

p-l- commented on scapy/layers/ipsec.py:
```diff
@@ -202,7 +198,13 @@ def __init__(self, name, cipher, mode, block_size=None, iv_size=None,
+            try:
+                self.is_aead = issubclass(self.mode,
+                                          modes.ModeWithAuthenticationTag)
+            except ImportError:
```
Are you certain this will raise `ImportError` and not an `AttibuteError`?

mtury replied:
Entirely right, sorry about that.

mtury commented:
@ThomasFaivre Indeed I ran some tests on the TLS module. I'd be mostly fine with a cryptography 1.2+ requirement. The only trouble would come from a FFDH `exchange()` which was added quite late, in version 1.7 (note that this version was packaged for Debian testing before the freeze, so that's good for us).

If @p-l- and @guedou are fine with this, I'll update the doc and put the appropriate warnings on the TLS side.

p-l- commented:
@mtury: would it be hard to (partly) support the version currently included with Debian stable?

mtury commented:
@p-l- If you mean version 0.6, it definitely would be difficult not to drop every crypto related methods. Many features were added to the library in the last two years, and several bugs were fixed.

If we want to ask for support of a version packaged in a Debian stable, version 1.7 will be packaged in the incoming stable (Stretch).

guedou commented:
@p-l- I believe that we can merge this patch and make 1.7 the mandatory python-cryptography version. 

@mtury could you prepare another PR that will display a warning if the cryptography version is too old ?

Note: Stretch is likely to be released this autumn.

mtury commented:
I renamed the PR because its content and the discussion was already more about `cryptography` than the IPsec module.

In order to raise warnings at the appropriate places without losing too much of scapy functionalities, I introduced a `crypto_valid` attribute in the global `conf`, along with a `crypto_validator` decorator.

Note that PR #476 already introduced some changes to `cert.py`... In order to avoid too much conflicts during merging, I'm waiting for the present PR to be accepted, and then I'll rework the TLS module.

guedou commented on scapy/config.py:
```diff
@@ -279,6 +279,14 @@ def __set__(self,obj,val):
         obj._logLevel = val
         
 
+def isCryptographyValid():
```
Could you add a docstring ?

guedou commented:
@mtury awesome PR ! could you add a docstring ? I will merge it ASAP.
